dashboard/src/shared/PackageRepositoriesService.ts

```diff
@@ -0,0 +1,233 @@
+        tlsCertKey: { ...request.tlsCertKey } as TlsCertKey,
+      } as PackageRepositoryAuth;
+    }
+    if (Object.values(request.opaqueCreds.data).some(e => !!e)) {
+      addPackageRepositoryRequest.auth = {
+        ...addPackageRepositoryRequest.auth,
+        opaqueCreds: { ...request.opaqueCreds } as OpaqueCredentials,
+      } as PackageRepositoryAuth;
+    }
+    if (request.customCA) {
+      addPackageRepositoryRequest.tlsConfig = {
+        ...addPackageRepositoryRequest.tlsConfig,
+        certAuthority: request.customCA,
+      } as PackageRepositoryTlsConfig;
+    }
+    if (request.skipTLS) {
+      addPackageRepositoryRequest.tlsConfig = {
+        ...addPackageRepositoryRequest.tlsConfig,
+        insecureSkipVerify: request.skipTLS,
+      } as PackageRepositoryTlsConfig;
+    }
+    if (request.secretTLSName) {
+      addPackageRepositoryRequest.tlsConfig = {
+        ...addPackageRepositoryRequest.tlsConfig,
+        secretRef: {
+          name: request.secretTLSName,
+        } as SecretKeyReference,
+      } as PackageRepositoryTlsConfig;
+    }
+    if (request.secretAuthName) {
+      addPackageRepositoryRequest.auth = {
+        ...addPackageRepositoryRequest.auth,
+        secretRef: {
+          name: request.secretAuthName,
+        } as SecretKeyReference,
+      } as PackageRepositoryAuth;
+    }
+
+    if (isUpdate) {
+      const updatePackageRepositoryRequest: UpdatePackageRepositoryRequest = {
+        description: addPackageRepositoryRequest.description,
+        interval: addPackageRepositoryRequest.interval,
+        url: addPackageRepositoryRequest.url,
+        auth: addPackageRepositoryRequest.auth,
+        customDetail: addPackageRepositoryRequest.customDetail,
+        tlsConfig: addPackageRepositoryRequest.tlsConfig,
+        packageRepoRef: {
+          identifier: addPackageRepositoryRequest.name,
+          context: addPackageRepositoryRequest.context,
+          plugin: addPackageRepositoryRequest.plugin,
+        },
+      };
+      return updatePackageRepositoryRequest;
+    }
+    return addPackageRepositoryRequest;
+  }
+
+  private static buildEncodedCustomDetails(request: IPkgRepoFormData) {
+    // if using the Helm plugin, add its custom fields.
+    // An "Any" object has  "typeUrl" with the FQN of the type and a "value",
+    // which is the result of the encoding (+finish(), to get the Uint8Array)
+    // of the actual custom object
+    switch (request.plugin?.name) {
+      case PluginNames.PACKAGES_HELM:
+        return {
+          typeUrl: `${helmProtobufPackage}.RepositoryCustomDetails`,
+          value: RepositoryCustomDetails.encode({
+            dockerRegistrySecrets: request.customDetails.dockerRegistrySecrets,
+            ociRepositories: request.customDetails.ociRepositories,
+            filterRule: request.customDetails.filterRule,
+            performValidation: request.customDetails.performValidation,
+          } as RepositoryCustomDetails).finish(),
```

Comment:
Just a suggestion, haven't tested it, but maybe we could do this?

```suggestion
          value: RepositoryCustomDetails.encode(request.customDetails as RepositoryCustomDetails).finish(),
```

Reply:
Having `any` as type instead of `Partial<RepositoryCustomDetails>` [here](https://github.com/vmware-tanzu/kubeapps/blob/c0598eb0411183723e927b23c8c2d6b6dcb19cb2/dashboard/src/shared/types.ts#L446) would be more generic, not depending on the Helm implementation. But again, havent't tested it. Perhaps for a future PR 😄

Reply:
Good idea, it should work. Testing it straight away, thanks!